### PR TITLE
Partial Sync: Support unsubscribe

### PR DESCRIPTION
**Pull request by cmelchior**
Makes it possible to cancel named subscriptions.

### CODE REVIEW

**Comment by cmelchior** [comment]
Ready for review @nhachicha

File: realm/realm-library/src/androidTestObjectServer/java/io/realm/SyncedRealmTests.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+            realm.unsubscribeAsync("my-id", callback);
+            fail();
+        } catch (UnsupportedOperationException ignore) {
+        } finally {
+            realm.close();
+        }
+
+        looperThread.testComplete();
+    }
+}
```
Comment (nhachicha) [inline realm/realm-library/src/androidTestObjectServer/java/io/realm/SyncedRealmTests.java]: add new line

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 2]
```diff
@@ -1697,6 +1700,70 @@ public static boolean compactRealm(RealmConfiguration configuration) {
         return BaseRealm.compactRealm(configuration);
     }
 
+    /**
+     * Cancel a named subscription that was created by calling
+     * {@link RealmQuery#findAllAsync(String)}. If after this, some objects are no longer part of
+     * any active  subscription they will be removed from the device (but not deleted on the server).
```
Comment (nhachicha) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: del extra space

Comment (nhachicha) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: they will be removed **locally** from the device?

Comment (nhachicha) [reply realm/realm-library/src/main/java/io/realm/Realm.java]: _stick to delete or remove in the same sentence_

but not **removed** on the server

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 3]
```diff
@@ -1697,6 +1700,70 @@ public static boolean compactRealm(RealmConfiguration configuration) {
...
+     *
+     * The effect of unsubscribing is not immediate. The local Realm must coordinate with the Object
+     * Server before this can happen. So a successful callback just indicate that the request was
```
Comment (nhachicha) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: del So

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/PartialSyncTests.java [hunk 4]
```diff
@@ -236,26 +208,84 @@ public void partialSync_namedSubscription_namedConflictThrows() {
...
+        });
+    }
+
+    @Test
+    @RunTestInLooperThread
+    public void unsubscribeAsync_nonExistingIdThrows() throws InterruptedException {
+        SyncUser user = UserFactory.createUniqueUser(Constants.AUTH_URL);
+        Realm realm = getPartialRealm(user);
+        looperThread.closeAfterTest(realm);
+
```
Comment (nhachicha) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/PartialSyncTests.java]: remove line

File: realm/realm-library/src/main/java/io/realm/Realm.java (hunk 2 above)
Comment (cmelchior) [reply realm/realm-library/src/main/java/io/realm/Realm.java]: I was actually going for a distinction between deleting (which is permanent) vs. remove (which is not), but maybe that is more confusing than helpful. I'll reword

**Comment by Zhuinden** [comment]
Supercedes https://github.com/realm/realm-java/pull/5538

**Pull request merged**